Add Prometheus metrics support to clustermesh-apiserver

giorio94 opened the pull request:
This PR extends the clustermesh-apiserver implementation and the corresponding Helm chart configuration to enable exposing and scraping Prometheus metrics. In addition to the metrics automatically exposed by etcd, the apiserver currently exposes kvstore-related metrics and basic go-related ones.

The correct functioning has been tested with a local kube-prometheus installation.

<!-- Description of change -->

```release-note
Add Prometheus metrics support to clustermesh-apiserver
```

giorio94 committed:
clustermesh-apiserver: add metrics server cell

This commit extends the clustermesh-apiserver with a new cell in
charge of exposing Prometheus metrics (disabled by default). It
currently exposes basic go-related metrics, and kvstore-related
metrics; Additional metrics will be introduced in subsequent
commits.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
clustermesh-apiserver: extend helm chart to expose metrics

This commit extends the helm chart to allow configuring the exposition
of Prometheus metrics for the clustermesh-apiserver component (including
both the apiserver and the etcd containers). Specifically, it adds the
corresponding configuration knobs and introduces a dedicated service and
servicemonitor (disabled by default).

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

jrajahalme commented on install/kubernetes/cilium/templates/clustermesh-apiserver/servicemonitor.yaml:
```diff
@@ -0,0 +1,57 @@
+    - {{ .Release.Namespace }}
+  endpoints:
+  {{- if .Values.clustermesh.apiserver.metrics.enabled }}
+  - port: apiserv-metrics
```
Any reason why these port names have a shorter form `apiserv-metrics` and  not `apiserver-metrics`?

giorio94 replied:
Yes, port names are limited to 15 characters to comply with IANA_SVC_NAME constrains: [RFC6335](https://www.rfc-archive.org/getrfc.php?rfc=6335#gsc.tab=0).

jrajahalme reviewed (approved):
LGTM. minor question on port naming.

gandro reviewed (approved):
Thanks!

giorio94 committed:
docs: add clustermesh-apiserver metrics

Extend the "Monitoring & Metrics" documentation page to include
information about the metrics exposed by the clustermesh-apiserver.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 commented on Documentation/observability/metrics.rst:
```diff
@@ -175,6 +175,39 @@ OpenMetrics imposes a few additional requirements on metrics names and labels,
+Cluster Mesh API Server metrics provide insights into both the state of the
+``clustermesh-apiserver`` process and the sidecar etcd instance.
+Cluster Mesh API Server metrics are exported under the ``cilium_`` Prometheus
+namespace. Etcd metrics are exported under the ``etcd_`` Prometheus namespace.
```
Are we fine with *Cluster Mesh API Server metrics are exported under the ``cilium_`` Prometheus
namespace*, or should we expose them under the `clustermesh_apiserver` namespace?

tommyp1ckles replied:
They don't use both, like the operator does (i.e. `cilium_clustermesh_apiserver_`)?

giorio94 replied:
Currently not, because this PR is exposing a few etcd metrics which were already present in cilium (using the same format). But I can make the `Namespace` part configurable and change it for the clustermesh-apiserver.

giorio94 replied:
Updated to `cilium_clustermesh_apiserver_`

qmonnet commented on Documentation/observability/metrics.rst:
```diff
@@ -855,3 +888,30 @@ Options
+Name                                     Labels                                       Description
+======================================== ============================================ ========================================================
+``kvstore_operations_duration_seconds``  ``action``, ``kind``, ``outcome``, ``scope`` Duration of kvstore operation
+``kvstore_events_queue_seconds``         ``action``, ``scope``                        Duration of seconds of time received event was blocked before it could be queued
```
I understand the description was copied from above in the doc, but it looks like it could be improved? `Time received event was blocked before it could be queued, in seconds`?

ChatGPT suggesting a few others :) 

![image](https://github.com/cilium/cilium/assets/17001771/a242be02-2f9e-40b6-b8d5-0d0db10fac80)

giorio94 replied:
Yeah, that description seemed a bit weird also to me. I've tried to rephrase it in a separate commit to *"Seconds waited before a received event was queued"*. Does that sound better to you? If you prefer any other alternative I'll update it.

qmonnet commented on Documentation/observability/metrics.rst:
```diff
@@ -175,6 +175,39 @@ OpenMetrics imposes a few additional requirements on metrics names and labels,
+
+You can automatically create a
+`Prometheus Operator <https://github.com/prometheus-operator/prometheus-operator>`_
+``ServiceMonitor`` setting ``clustermesh.apiserver.metrics.serviceMonitor.enabled=true``.
```
```suggestion
``ServiceMonitor`` by setting ``clustermesh.apiserver.metrics.serviceMonitor.enabled=true``.
```

qmonnet reviewed (approved):
I looked mostly at the docs - They look good, here are a few minor suggestions below.

giorio94 committed:
metrics: rephrase description of kvstore_events_queue_seconds metric

Rephrase the description of the 'kvstore_events_queue_seconds` metrics
for improved clarity.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
metrics: make the metrics namespace configurable

Currently, the metrics namespace is hard-coded to the `cilium` value.
This commit changes it to be a variable (assigned by default the same
value) to allow changing it when metrics are exposed by a different
component (e.g., clustermesh-apiserver).

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>
